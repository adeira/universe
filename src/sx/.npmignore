__flowtests__
__tests__
