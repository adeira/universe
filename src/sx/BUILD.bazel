load("//:bazel/macros/build_yarn_workspace.bzl", "build_yarn_workspace")
load("//:bazel/macros/test_yarn_workspace.bzl", "test_yarn_workspace")

build_yarn_workspace(
    name = "sx",
    package_name = "@adeira/sx",
)

test_yarn_workspace(
    name = "test",
    deps = [
        "@npm//@testing-library/react",
        "@npm//change-case",
        "@npm//css-tree",
        "@npm//fast-levenshtein",
        "@npm//prettier",
        "@npm//react-test-renderer",
        "@universe//src/fixtures-tester",
        "@universe//src/js",
        "@universe//src/murmur-hash",
    ],
)
