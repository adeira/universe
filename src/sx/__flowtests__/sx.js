// @flow

import React, { type Node } from 'react';

import * as sx from '../index';

export function NoErrors(): Node {
  const styles = sx.create({
    red: { color: 'red' },
  });
  return <div className={styles('red')} />;
}

export function NoErrorsWithPseudo(): Node {
  const styles = sx.create({
    red: {
      'color': 'red',
      ':hover': {
        color: 'blue',
      },
      '::after': {
        content: '"â˜…"',
      },
    },
  });
  return <div className={styles('red')}>I am blue on hover</div>;
}

export function NoErrorsMoreStylesheetNames(): Node {
  const styles = sx.create({
    red: { color: 'red' },
    blue: { color: 'blue' },
  });
  return <div className={styles('red', 'blue')} />;
}

export function InvalidStylesheetName(): Node {
  const styles = sx.create({
    red: { color: 'red' },
  });
  // $FlowExpectedError[incompatible-call]
  return <div className={styles('invalid')} />;
}

export function InvalidCSSProperty(): Node {
  const styles = sx.create({
    red: {
      // $FlowExpectedError[incompatible-call]
      wtf: 'wtf',
    },
  });
  // $FlowExpectedError[incompatible-call]
  return <div className={styles('invalid')} />;
}

export function InvalidCSSPseudoProperty(): Node {
  const styles = sx.create({
    red: {
      // $FlowExpectedError[incompatible-call]
      ':hover': -1,
    },
  });
  return <div className={styles('red')} />;
}

export function FlexNonNumericValues(): Node {
  const styles = sx.create({
    'flex-1': { flex: '1 1 0%' },
    'flex-auto': { flex: '1 1 auto' },
    'flex-initial': { flex: '0 1 auto' },
    'flex-none': { flex: 'none' },
  });
  return <div className={styles('flex-1', 'flex-auto', 'flex-initial', 'flex-none')} />;
}

export function GridColumnNonNumericValues(): Node {
  const styles = sx.create({
    'col-auto': { gridColumn: 'auto' },
    'col-end-auto': { gridColumnEnd: 'auto' },
    'col-span-1': { gridColumn: 'span 1 / span 1' },
  });
  return <div className={styles('col-auto', 'col-end-auto', 'col-span-1')} />;
}

export function GridRowNonNumericValues(): Node {
  const styles = sx.create({
    'row-auto': { gridRow: 'auto' },
    'row-end-auto': { gridRowEnd: 'auto' },
    'row-span-1': { gridRow: 'span 1 / span 1' },
    'row-start-auto': { gridRowStart: 'auto' },
  });
  return <div className={styles('row-auto', 'row-end-auto', 'row-span-1', 'row-start-auto')} />;
}

export function ZIndex(): Node {
  const styles = sx.create({
    'z-auto': { zIndex: 'auto' },
    'z-10': { zIndex: 10 },
  });
  return <div className={styles('z-auto', 'z-10')} />;
}
