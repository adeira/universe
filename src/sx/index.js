/**
 * Users of SX should be able to import it as follows:
 *
 *    import sx from '@adeira/sx';
 *    sx(), sx.create(), sx.keyframes(), ...
 *
 * Alternatively:
 *
 *    import { create as sxCreate } from '@adeira/sx';
 *    sxCreate(), ...
 *
 * @flow
 */

import create, { type AllCSSProperties } from './src/create';
import keyframes from './src/keyframes';
import renderPageWithSX from './src/renderPageWithSX';

/**
 * This function allows us to compose local and external styles like so:
 *
 * ```
 * import sx from '@adeira/sx';
 *
 * const styles = sx.create({ default: { fontSize: 16 } });
 * const externalStyles = sx.create({ spacing: { marginTop: 4 } });
 *
 * <div className={sx(styles.default, externalStyles.spacing)} />;
 * ```
 *
 * It does so by literally merging the 2 objects together and calling `sx.create` on the result.
 */
function composeStylesheets(...stylesheets: $ReadOnlyArray<?AllCSSProperties>): ?string {
  // Should we support deeply nested styles or leave it like this and overwrite them?
  const mergedStylesheet = Object.assign({}, ...stylesheets);
  if (Object.keys(mergedStylesheet).length === 0) {
    // happens when composing "nothing" which is a valid input
    return undefined;
  }
  return create({
    __internal_compose_stylesheets: mergedStylesheet,
  })('__internal_compose_stylesheets');
}

composeStylesheets.create = create;
composeStylesheets.keyframes = keyframes;
composeStylesheets.renderPageWithSX = renderPageWithSX;

export { create, keyframes, renderPageWithSX };
export default composeStylesheets;
export type { AllCSSProperties } from './src/create';
