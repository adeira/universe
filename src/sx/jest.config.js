// @flow

const path = require('path');

module.exports = {
  rootDir: __dirname,
  setupFilesAfterEnv: [(path.join(__dirname, 'scripts', 'setupTests.js') /*: string */)],
};
