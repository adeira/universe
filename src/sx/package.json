{
  "name": "@adeira/sx",
  "description": "Atomic CSS-in-JS (not only) for Next.js applications",
  "homepage": "https://github.com/adeira/universe/tree/master/src/sx",
  "license": "MIT",
  "private": false,
  "version": "0.27.0",
  "main": "./index.js",
  "type": "commonjs",
  "sideEffects": false,
  "dependencies": {
    "@adeira/css-colors": "^2.0.0",
    "@adeira/js": "^2.1.0",
    "@adeira/murmur-hash": "^2.0.0",
    "@adeira/signed-source": "^2.0.0",
    "@babel/runtime": "^7.14.6",
    "change-case": "^4.1.2",
    "css-tree": "^1.1.3",
    "fast-levenshtein": "^3.0.0",
    "json-stable-stringify": "^1.0.1",
    "mdn-data": "^2.0.19",
    "prettier": "^2.3.2",
    "stylis": "^4.0.10"
  },
  "peerDependencies": {
    "react": "^17.0.0"
  },
  "devDependencies": {
    "@adeira/fixtures-tester": "^1.1.0",
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^12.0.0",
    "react": "^17.0.2",
    "react-test-renderer": "^17.0.2",
    "snapshot-diff": "^0.9.0"
  }
}
