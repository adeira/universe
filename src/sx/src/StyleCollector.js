// @flow

import { invariant, warning } from '@adeira/js';

import StyleCollectorAtNode from './StyleCollectorAtNode';
import StyleCollectorNode, { type StyleCollectorNodeInterface } from './StyleCollectorNode';
import StyleCollectorPseudoNode from './StyleCollectorPseudoNode';

// "hashRegistry": Map {
//   "aaa" => Map {
//     "color" => "c0",
//   },
//   "bbb" => Map {
//     "fontSize" => "c1",
//     "color" => "c2",
//   },
// },
type HashRegistryType = Map<string, Map<string, string>>;

// "styleBuffer": Map {
//   "c0" => StyleCollectorNode {
//     "hash": "c0",
//     "styleName": "color",
//     "styleValue": "#00f",
//   },
//   "c1" => StyleCollectorNode {
//     "hash": "c1",
//     "styleName": "font-size",
//     "styleValue": "1rem",
//   },
//   "c2" => StyleCollectorNode {
//     "hash": "c2",
//     "styleName": "color",
//     "styleValue": "#00f",
//   },
// },
type StyleBufferType = Map<string, StyleCollectorNodeInterface>;

class StyleCollector {
  #styleBuffer: StyleBufferType = new Map();

  collect(baseStyleSheet: {|
    +[sheetName: string]: $FlowFixMe,
  |}): {| +hashRegistry: HashRegistryType, +styleBuffer: StyleBufferType |} {
    const hashRegistry: HashRegistryType = new Map();

    const traverse = (styleSheetName, styleSheetObject, styleBuffer, hashSeed = '') => {
      // we need to iterate it - there might be pseudo classes, at rules or leaf types
      for (const maybeName of Object.keys(styleSheetObject)) {
        const maybeValue = styleSheetObject[maybeName];
        if (typeof maybeValue === 'number' || typeof maybeValue === 'string') {
          // basic leaf type
          const node = new StyleCollectorNode(maybeName, maybeValue, hashSeed);
          styleBuffer.set(node.getHash(), node);
          // add record to the hash registry
          const hashRegistryKey = `${maybeName}${hashSeed}`;
          if (hashRegistry.has(styleSheetName)) {
            hashRegistry.get(styleSheetName)?.set(hashRegistryKey, node.getHash());
          } else {
            hashRegistry.set(styleSheetName, new Map([[hashRegistryKey, node.getHash()]]));
          }
        } else if (maybeName.startsWith(':')) {
          // pseudo type (:hover, ::after, ...)
          for (const key of Object.keys(maybeValue)) {
            invariant(key.startsWith(':') === false, 'Nested pseudo classes are not allowed.');
          }
          const nodes = traverse(styleSheetName, maybeValue, new Map(), maybeName);
          if (styleBuffer.has(maybeName)) {
            styleBuffer.get(maybeName)?.addNodes(nodes);
          } else {
            const node = new StyleCollectorPseudoNode(maybeName, nodes);
            styleBuffer.set(node.getPseudo(), node);
          }
        } else if (maybeName.startsWith('@media') || maybeName.startsWith('@supports')) {
          // at rule type (@media, @supports, ...)
          const nodes = traverse(styleSheetName, maybeValue, new Map(), maybeName);
          if (styleBuffer.has(maybeName)) {
            styleBuffer.get(maybeName)?.addNodes(nodes);
          } else {
            const node = new StyleCollectorAtNode(maybeName, nodes);
            styleBuffer.set(node.getAtRuleName(), node);
          }
        } else {
          // be silent in production (warning only)
          const messsage = `Unsupported rule "%s"`;
          invariant(!__DEV__, messsage, maybeName);
          warning(__DEV__, messsage, maybeName);
        }
      }
      return styleBuffer;
    };

    for (const styleSheetName of Object.keys(baseStyleSheet)) {
      const styleSheetObject = baseStyleSheet[styleSheetName];
      traverse(styleSheetName, styleSheetObject, this.#styleBuffer);
    }

    return {
      styleBuffer: this.#styleBuffer,
      hashRegistry: hashRegistry,
    };
  }

  print(): string {
    let sxStyle = '';
    this.#styleBuffer.forEach((node) => {
      sxStyle += node.print();
    });
    return sxStyle;
  }

  reset(): void {
    this.#styleBuffer.clear();
  }
}

const styleCollector = new StyleCollector();

export default (styleCollector: StyleCollector);
