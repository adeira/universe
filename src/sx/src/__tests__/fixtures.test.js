// @flow

import fs from 'fs';
import os from 'os';
import path from 'path';
import prettier from 'prettier';
import TestRenderer from 'react-test-renderer';
import { isObject, sprintf } from '@adeira/js';

import sx from '../../index';
import StyleCollector from '../StyleCollector';

const fixturesPath = path.join(__dirname, 'fixtures');

expect.addSnapshotSerializer({
  print(serializerValue) {
    return Object.keys(serializerValue)
      .map((key) => {
        const value = serializerValue[key];
        const inspectedValue = isObject(value) ? JSON.stringify(value, null, 2) : value;
        return `~~~~~~~~~~ ${key.toUpperCase()} ~~~~~~~~~~\n${inspectedValue}`;
      })
      .join(os.EOL);
  },
  test() {
    return true;
  },
});

const renderPageMock = () => ({
  html: '',
  head: [''],
  styles: [''],
});

afterEach(() => {
  StyleCollector.reset();
});

const fixturesPaths = fs.readdirSync(fixturesPath);

test.each(fixturesPaths)('matches expected output: %s', (fixturePath) => {
  // 1) original definitions
  const stylesheetsDefinition = require(path.join(fixturesPath, fixturePath)).default;

  try {
    const styles = sx.create(stylesheetsDefinition);
    // 2) atomic CSS
    const renderer = TestRenderer.create(sx.renderPageWithSX(renderPageMock).styles[0]);

    const css = renderer.root.props.dangerouslySetInnerHTML.__html;
    const output = prettier.format(css, { filepath: 'test.css' });

    // 3) usage resulting CSS class names
    let usage = '';
    Object.keys(stylesheetsDefinition).forEach((stylesheetName) => {
      usage += sprintf(
        `
className={styles('%s')}
  ↓ ↓ ↓
class="%s"
`,
        stylesheetName,
        styles(stylesheetName),
      );
    });

    expect({
      input: stylesheetsDefinition,
      output,
      usage,
    }).toMatchSnapshot();
  } catch (error) {
    expect({
      input: stylesheetsDefinition,
      output: `⚠️ THROWN EXCEPTION:\n\n${error.toString()}`,
    }).toMatchSnapshot();
  }
});
