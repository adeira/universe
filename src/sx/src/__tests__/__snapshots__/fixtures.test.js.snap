// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: @supports.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "test": {
    "@supports (display: grid)": {
      "display": "grid"
    },
    "@supports not (display: grid)": {
      "float": "right"
    }
  },
  "testComplex": {
    "@supports (display: table-cell) and ((display: list-item) and (display:run-in))": {
      "color": "red"
    },
    "@supports (transform-style: preserve) or (-moz-transform-style: preserve) or (-o-transform-style: preserve) or (-webkit-transform-style: preserve)": {
      "color": "red"
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
@supports (display: grid) {
  ._3GKV06._3GKV06 {
    display: -ms-grid;
    display: grid;
  }
}
@supports not (display: grid) {
  ._1cjV2G._1cjV2G {
    float: right;
  }
}
@supports (display: table-cell) and ((display: list-item) and (display: run-in)) {
  ._3u8TFK._3u8TFK {
    color: #f00;
  }
}
@supports (transform-style: preserve) or (-moz-transform-style: preserve) or
  (-o-transform-style: preserve) or (-webkit-transform-style: preserve) {
  ._2gYAIc._2gYAIc {
    color: #f00;
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('test')}
  ↓ ↓ ↓
class="_3GKV06 _1cjV2G"

className={styles('testComplex')}
  ↓ ↓ ↓
class="_3u8TFK _2gYAIc"

`;

exports[`matches expected output: @supports-nested.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "test": {
    "@supports (display: grid)": {
      "color": "red",
      "@supports not (display: grid)": {
        "color": "blue",
        "@supports (display: table-cell) and ((display: list-item) and (display:run-in))": {
          "color": "orange"
        }
      }
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
@supports (display: grid) {
  ._4gBjL3._4gBjL3 {
    color: #f00;
  }
  @supports not (display: grid) {
    ._2HFxQl._2HFxQl {
      color: #00f;
    }
    @supports (display: table-cell) and
      ((display: list-item) and (display: run-in)) {
      ._1gIn7I._1gIn7I {
        color: #ffa500;
      }
    }
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('test')}
  ↓ ↓ ↓
class="_4gBjL3 _2HFxQl _1gIn7I"

`;

exports[`matches expected output: @unknown.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "test": {
    "@unknown": {
      "color": "red"
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
⚠️ THROWN EXCEPTION:

Invariant Violation: Unsupported rule "@unknown"
`;

exports[`matches expected output: autoprefix.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "title": {
    "backgroundClip": "text"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._2D4soO {
  -webkit-background-clip: text;
  background-clip: text;
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('title')}
  ↓ ↓ ↓
class="_2D4soO"

`;

exports[`matches expected output: keyframes-from-to.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "aaa": {
    "animationName": "_2rMlJa"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
.P4y5l {
  -webkit-animation-name: _2rMlJa;
  animation-name: _2rMlJa;
}
@-webkit-keyframes _2rMlJa {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
@keyframes _2rMlJa {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('aaa')}
  ↓ ↓ ↓
class="P4y5l"

`;

exports[`matches expected output: keyframes-percentage.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "aaa": {
    "animationName": "a2yPF"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._3SIQcJ {
  -webkit-animation-name: a2yPF;
  animation-name: a2yPF;
}
@-webkit-keyframes a2yPF {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
  25% {
    opacity: 0.7;
  }
}
@keyframes a2yPF {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
  25% {
    opacity: 0.7;
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('aaa')}
  ↓ ↓ ↓
class="_3SIQcJ"

`;

exports[`matches expected output: media-queries.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "test": {
    "color": "red",
    "@media print": {
      "color": "red"
    },
    "@media (min-width: 30em) and (max-width: 50em)": {
      "color": "blue"
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._324Crd {
  color: #f00;
}
@media print {
  .O347H.O347H {
    color: #f00;
  }
}
@media (min-width: 30em) and (max-width: 50em) {
  ._3yLn7C._3yLn7C {
    color: #00f;
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('test')}
  ↓ ↓ ↓
class="_324Crd O347H _3yLn7C"

`;

exports[`matches expected output: media-queries-multiple.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "aaa": {
    "color": "red",
    "fontSize": 32,
    "@media print": {
      "color": "red",
      "fontSize": 16
    },
    "@media (min-width: 30em) and (max-width: 50em)": {
      "color": "blue"
    }
  },
  "bbb": {
    "@media print": {
      "color": "green"
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._324Crd {
  color: #f00;
}
._1DKsqE {
  font-size: 32px;
}
@media print {
  .O347H.O347H {
    color: #f00;
  }
  .gdWvu.gdWvu {
    font-size: 16px;
  }
  ._1Yuw4C._1Yuw4C {
    color: #008000;
  }
}
@media (min-width: 30em) and (max-width: 50em) {
  ._3yLn7C._3yLn7C {
    color: #00f;
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('aaa')}
  ↓ ↓ ↓
class="_324Crd _1DKsqE O347H gdWvu _3yLn7C"

className={styles('bbb')}
  ↓ ↓ ↓
class="_1Yuw4C"

`;

exports[`matches expected output: media-queries-nested.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "test": {
    "@media print": {
      "color": "red",
      "@media (max-width: 12cm)": {
        "color": "blue"
      }
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
@media print {
  .O347H.O347H {
    color: #f00;
  }
  @media (max-width: 12cm) {
    ._3XWcpy._3XWcpy {
      color: #00f;
    }
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('test')}
  ↓ ↓ ↓
class="O347H _3XWcpy"

`;

exports[`matches expected output: media-queries-nested-deep.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "test": {
    "@media (max-width: 10cm)": {
      "color": "#F00000",
      "backgroundColor": "#FFFFFF",
      "@media (max-width: 9cm)": {
        "color": "#0F0000",
        "backgroundColor": "#FFFFFF",
        "@media (max-width: 8cm)": {
          "color": "#00F000",
          "backgroundColor": "#FFFFFF",
          "@media (max-width: 7cm)": {
            "color": "#000F00",
            "backgroundColor": "#FFFFFF",
            "@media (max-width: 6cm)": {
              "color": "#0000F0",
              "backgroundColor": "#FFFFFF",
              "@media (max-width: 5cm)": {
                "color": "#00000F",
                "backgroundColor": "#FFFFFF"
              }
            }
          }
        }
      }
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
@media (max-width: 10cm) {
  ._2Y4Iwc._2Y4Iwc {
    color: #f00000;
  }
  ._4yjauT._4yjauT {
    background-color: #fff;
  }
  @media (max-width: 9cm) {
    ._33Zsp._33Zsp {
      color: #0f0000;
    }
    ._356q9B._356q9B {
      background-color: #fff;
    }
    @media (max-width: 8cm) {
      ._3e6rQn._3e6rQn {
        color: #00f000;
      }
      ._2G8HFc._2G8HFc {
        background-color: #fff;
      }
      @media (max-width: 7cm) {
        ._5cFAt._5cFAt {
          color: #000f00;
        }
        ._1uNKjr._1uNKjr {
          background-color: #fff;
        }
        @media (max-width: 6cm) {
          ._3Y2upP._3Y2upP {
            color: #0000f0;
          }
          ._1j0v0B._1j0v0B {
            background-color: #fff;
          }
          @media (max-width: 5cm) {
            ._1yZ6Vs._1yZ6Vs {
              color: #00000f;
            }
            ._1gfVcg._1gfVcg {
              background-color: #fff;
            }
          }
        }
      }
    }
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('test')}
  ↓ ↓ ↓
class="_2Y4Iwc _4yjauT _33Zsp _356q9B _3e6rQn _2G8HFc _5cFAt _1uNKjr _3Y2upP _1j0v0B _1yZ6Vs _1gfVcg"

`;

exports[`matches expected output: media-queries-nested-pseudo.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "media": {
    "color": "red",
    "@media print": {
      "color": "red",
      ":hover": {
        "color": "pink"
      }
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._324Crd {
  color: #f00;
}
@media print {
  .O347H.O347H {
    color: #f00;
  }
  ._2sykgO._2sykgO:hover {
    color: #ffc0cb;
  }
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('media')}
  ↓ ↓ ↓
class="_324Crd O347H _2sykgO"

`;

exports[`matches expected output: media-queries-same-style.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "style1": {
    "@media (min-width: 1280px)": {
      "color": "red"
    },
    "color": "blue"
  },
  "style2": {
    "color": "red"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
@media (min-width: 1280px) {
  ._4k1sZK._4k1sZK {
    color: #f00;
  }
}
._2dHaKY {
  color: #00f;
}
._324Crd {
  color: #f00;
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('style1')}
  ↓ ↓ ↓
class="_4k1sZK _2dHaKY"

className={styles('style2')}
  ↓ ↓ ↓
class="_324Crd"

`;

exports[`matches expected output: pseudo-classes.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "link": {
    "textDecoration": "none",
    ":hover": {
      "textDecoration": "underline"
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
.v2kHO {
  -webkit-text-decoration: none;
  text-decoration: none;
}
.crve5:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('link')}
  ↓ ↓ ↓
class="v2kHO crve5"

`;

exports[`matches expected output: pseudo-classes-multiple.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "aaa": {
    "textDecoration": "none",
    ":hover": {
      "textDecoration": "none",
      "color": "red"
    },
    ":active": {
      "color": "red"
    }
  },
  "bbb": {
    ":hover": {
      "color": "blue"
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
.v2kHO {
  -webkit-text-decoration: none;
  text-decoration: none;
}
._30RBbb:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}
._1O0igU:hover {
  color: #f00;
}
._3nDIb1:hover {
  color: #00f;
}
._2kD1Fb:active {
  color: #f00;
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('aaa')}
  ↓ ↓ ↓
class="v2kHO _30RBbb _1O0igU _2kD1Fb"

className={styles('bbb')}
  ↓ ↓ ↓
class="_3nDIb1"

`;

exports[`matches expected output: pseudo-classes-nested-1.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "test": {
    ":hover": {
      "color": "red",
      ":hover": {
        "color": "green",
        ":hover": {
          "color": "blue"
        }
      }
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
⚠️ THROWN EXCEPTION:

Invariant Violation: Nested pseudo classes are not allowed.
`;

exports[`matches expected output: pseudo-classes-nested-2.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "test": {
    ":hover::after": {
      "content": "\\"this is OK\\""
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._4yRZ3O:hover::after {
  content: "this is OK";
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('test')}
  ↓ ↓ ↓
class="_4yRZ3O"

`;

exports[`matches expected output: pseudo-elements.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "link": {
    "textDecoration": "none",
    ":hover": {
      "textDecoration": "underline"
    },
    "::after": {
      "content": "\\"∞\\""
    },
    "::before": {
      "content": "\\"∞\\""
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
.v2kHO {
  -webkit-text-decoration: none;
  text-decoration: none;
}
.crve5:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}
._2sj7wN::after {
  content: "∞";
}
._4vATsH::before {
  content: "∞";
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('link')}
  ↓ ↓ ↓
class="v2kHO crve5 _2sj7wN _4vATsH"

`;

exports[`matches expected output: pseudo-elements-2.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "ul": {
    "counterReset": "my-reset"
  },
  "li": {
    "counterIncrement": "my-reset",
    "::before": {
      "content": "counter(my-reset)"
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._33076O {
  counter-reset: my-reset;
}
._3DgpOy {
  counter-increment: my-reset;
}
._3ybBud::before {
  content: counter(my-reset);
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('ul')}
  ↓ ↓ ↓
class="_33076O"

className={styles('li')}
  ↓ ↓ ↓
class="_3DgpOy _3ybBud"

`;

exports[`matches expected output: readme-example.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "example": {
    "fontSize": 32,
    "textDecoration": "none",
    "backgroundColor": "var(--main-bg-color)"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._1DKsqE {
  font-size: 32px;
}
.v2kHO {
  -webkit-text-decoration: none;
  text-decoration: none;
}
.stDQH {
  background-color: var(--main-bg-color);
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('example')}
  ↓ ↓ ↓
class="_1DKsqE v2kHO stDQH"

`;

exports[`matches expected output: simple.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "redStyle": {
    "color": "red"
  },
  "blueStyle": {
    "color": "blue"
  },
  "greenStyle": {
    "color": "green"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._324Crd {
  color: #f00;
}
._2dHaKY {
  color: #00f;
}
.mRoJ3 {
  color: #008000;
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('redStyle')}
  ↓ ↓ ↓
class="_324Crd"

className={styles('blueStyle')}
  ↓ ↓ ↓
class="_2dHaKY"

className={styles('greenStyle')}
  ↓ ↓ ↓
class="mRoJ3"

`;

exports[`matches expected output: simple-transforms.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "red_one": {
    "color": "red"
  },
  "red_two": {
    "color": "#f00"
  },
  "red_three": {
    "color": "#FF0000"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._324Crd {
  color: #f00;
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('red_one')}
  ↓ ↓ ↓
class="_324Crd"

className={styles('red_two')}
  ↓ ↓ ↓
class="_324Crd"

className={styles('red_three')}
  ↓ ↓ ↓
class="_324Crd"

`;

exports[`matches expected output: simple-with-duplicates.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  "aaa": {
    "color": "red"
  },
  "bbb": {
    "color": "red"
  },
  "ccc": {
    "color": "red"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._324Crd {
  color: #f00;
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles('aaa')}
  ↓ ↓ ↓
class="_324Crd"

className={styles('bbb')}
  ↓ ↓ ↓
class="_324Crd"

className={styles('ccc')}
  ↓ ↓ ↓
class="_324Crd"

`;

exports[`matches expected output: variables.js 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
{
  ":root": {
    "--main-bg-color": "pink",
    "--someCOMPLEXvalue": "3px 6px rgb(20, 32, 54)"
  },
  "body": {
    "backgroundColor": "var(--main-bg-color)"
  },
  "bodyWithFallback": {
    "backgroundColor": "var(--someCOMPLEXvalue, blue)"
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
._4qI4JK {
  --main-bg-color: #ffc0cb;
}
._3zBmgU {
  --someCOMPLEXvalue: 3px 6px rgb(20, 32, 54);
}
.stDQH {
  background-color: var(--main-bg-color);
}
.S1vUl {
  background-color: var(--someCOMPLEXvalue, #00f);
}

~~~~~~~~~~ USAGE ~~~~~~~~~~

className={styles(':root')}
  ↓ ↓ ↓
class="_4qI4JK _3zBmgU"

className={styles('body')}
  ↓ ↓ ↓
class="stDQH"

className={styles('bodyWithFallback')}
  ↓ ↓ ↓
class="S1vUl"

`;
